# Automatidata project 
**Course 4 - The Power of Statistics**

You are a data professional in a data consulting firm, called Automatidata. The current project for their newest client, the New York City Taxi & Limousine Commission (New York City TLC) is reaching its midpoint, having completed a project proposal, Python coding work, and exploratory data analysis.

You receive a new email from Uli King, Automatidata’s project manager. Uli tells your team about a new request from the New York City TLC: to analyze the relationship between fare amount and payment type. A follow-up email from Luana includes your specific assignment: to conduct an A/B test. 

A notebook was structured and prepared to help you in this project. Please complete the following questions.


# Course 4 End-of-course project: Statistical analysis

In this activity, you will practice using statistics to analyze and interpret data. The activity covers fundamental concepts such as descriptive statistics and hypothesis testing. You will explore the data provided and conduct A/B and hypothesis testing.  
<br/>   

**The purpose** of this project is to demostrate knowledge of how to prepare, create, and analyze A/B tests. Your A/B test results should aim to find ways to generate more revenue for taxi cab drivers.

**Note:** For the purpose of this exercise, assume that the sample data comes from an experiment in which customers are randomly selected and divided into two groups: 1) customers who are required to pay with credit card, 2) customers who are required to pay with cash. Without this assumption, we cannot draw causal conclusions about how payment method affects fare amount.

**The goal** is to apply descriptive statistics and hypothesis testing in Python. The goal for this A/B test is to sample data and analyze whether there is a relationship between payment type and fare amount. For example: discover if customers who use credit cards pay higher fare amounts than customers who use cash.
  
*This activity has four parts:*

**Part 1:** Imports and data loading
* What data packages will be necessary for hypothesis testing?

**Part 2:** Conduct EDA and hypothesis testing
* How did computing descriptive statistics help you analyze your data? 

* How did you formulate your null hypothesis and alternative hypothesis? 

**Part 3:** Communicate insights with stakeholders

* What key business insight(s) emerged from your A/B test?

* What business recommendations do you propose based on your results?

<br/> 
Follow the instructions and answer the questions below to complete the activity. Then, you will complete an Executive Summary using the questions listed on the PACE Strategy Document.

Be sure to complete this activity before moving on. The next course item will provide you with a completed exemplar to compare to your own work. 

# **Conduct an A/B test**


<img src="images/Pace.png" width="100" height="100" align=left>

# **PACE stages**


Throughout these project notebooks, you'll see references to the problem-solving framework PACE. The following notebook components are labeled with the respective PACE stage: Plan, Analyze, Construct, and Execute.

<img src="images/Plan.png" width="100" height="100" align=left>


## PACE: Plan 

In this stage, consider the following questions where applicable to complete your code response:
1. What is your research question for this data project? Later on, you will need to formulate the null and alternative hypotheses as the first step of your hypothesis test. Consider your research question now, at the start of this task.


What is the relationship between fare amount and payment type?

*Complete the following steps to perform statistical analysis of your data:* 

### Task 1. Imports and data loading

Import packages and libraries needed to compute descriptive statistics and conduct a hypothesis test.

<details>
  <summary><h4><strong>Hint: </strong></h4></summary>

Before you begin, recall the following Python packages and functions that may be useful:

*Main functions*: stats.ttest_ind(a, b, equal_var)

*Other functions*: mean() 

*Packages*: pandas, stats.scipy

</details>

In [1]:
#==> ENTER YOUR CODE HERE
import pandas as pd
from scipy import stats

**Note:** As shown in this cell, the dataset has been automatically loaded in for you. You do not need to download the .csv file, or provide more code, in order to access the dataset and proceed with this lab. Please continue with this activity by completing the following instructions.

In [2]:
# Load dataset into dataframe
taxi_data = pd.read_csv("2017_Yellow_Taxi_Trip_Data.csv", index_col = 0)

<img src="images/Analyze.png" width="100" height="100" align=left>

<img src="images/Construct.png" width="100" height="100" align=left>

## PACE: **Analyze and Construct**

In this stage, consider the following questions where applicable to complete your code response:
1. Data professionals use descriptive statistics for Exploratory Data Analysis. How can computing descriptive statistics help you learn more about your data in this stage of your analysis?


It gives you a view of what the data is pointing out which is in turn is used to make further predictions or decisions.

### Task 2. Data exploration

Use descriptive statistics to conduct Exploratory Data Analysis (EDA). 

<details>
  <summary><h4><strong>Hint: </strong></h4></summary>

Refer back to *Self Review Descriptive Statistics* for this step-by-step proccess.

</details>

**Note:** In the dataset, `payment_type` is encoded in integers:
*   1: Credit card
*   2: Cash
*   3: No charge
*   4: Dispute
*   5: Unknown



In [3]:
#==> ENTER YOUR CODE HERE
print('data description')
print(taxi_data.describe())

print('/n')

print('first 5 rows of sample data')
print(taxi_data.head())

print('/n')
print(taxi_data.groupby('payment_type')['fare_amount'].sum())

data description
           VendorID  passenger_count  trip_distance    RatecodeID  \
count  22699.000000     22699.000000   22699.000000  22699.000000   
mean       1.556236         1.642319       2.913313      1.043394   
std        0.496838         1.285231       3.653171      0.708391   
min        1.000000         0.000000       0.000000      1.000000   
25%        1.000000         1.000000       0.990000      1.000000   
50%        2.000000         1.000000       1.610000      1.000000   
75%        2.000000         2.000000       3.060000      1.000000   
max        2.000000         6.000000      33.960000     99.000000   

       PULocationID  DOLocationID  payment_type   fare_amount         extra  \
count  22699.000000  22699.000000  22699.000000  22699.000000  22699.000000   
mean     162.412353    161.527997      1.336887     13.026629      0.333275   
std       66.633373     70.139691      0.496211     13.243791      0.463097   
min        1.000000      1.000000      1.0000

You are interested in the relationship between payment type and the fare amount the customer pays. One approach is to look at the average fare amount for each payment type. 

In [4]:
#==> ENTER YOUR CODE HERE
avg_fare_amount_byType = taxi_data.groupby('payment_type')['fare_amount'].mean()
avg_fare_amount_byType

payment_type
1    13.429748
2    12.213546
3    12.186116
4     9.913043
Name: fare_amount, dtype: float64

Based on the averages shown, it appears that customers who pay in credit card tend to pay a larger fare amount than customers who pay in cash. However, this difference might arise from random sampling, rather than being a true difference in fare amount. To assess whether the difference is statistically significant, you conduct a hypothesis test.


### Task 3. Hypothesis testing

Before you conduct your hypothesis test, consider the following questions where applicable to complete your code response:

1. Recall the difference between the null hypothesis and the alternative hypotheses. Consider your hypotheses for this project as listed below.

$H_0$: There is no difference in the average fare amount between customers who use credit cards and customers who use cash.

$H_A$: There is a difference in the average fare amount between customers who use credit cards and customers who use cash.



Your goal in this step is to conduct a two-sample t-test. Recall the steps for conducting a hypothesis test: 


1.   State the null hypothesis and the alternative hypothesis
2.   Choose a signficance level
3.   Find the p-value
4.   Reject or fail to reject the null hypothesis 



**Note:** For the purpose of this exercise, your hypothesis test is the main component of your A/B test. 

You choose 5% as the significance level and proceed with a two-sample t-test.

In [5]:
# 1. Stating the null hypothesis and the alternative hypothesis

#==> ENTER YOUR CODE HERE
# For credit card
selected_columns = ['payment_type', 'fare_amount']
avgFare_amount_credit = taxi_data[taxi_data['payment_type'] == 1][selected_columns]
avgFare_amount_credit

payment_type  fare_amount
24870114              1         13.0
35634249              1         16.0
106203690             1          6.5
38942136              1         20.5
23345809              1          9.0
...                 ...          ...
60425673              1         19.0
67858616              1          7.5
66632549              1         52.0
60217333              1         10.5
17208911              1         11.0

[15265 rows x 2 columns]

In [6]:
# for cash
avgFare_amount_cash = taxi_data[taxi_data['payment_type']==2][['payment_type', 'fare_amount']]
avgFare_amount_cash

payment_type  fare_amount
30841670              2         16.5
8433159               2          9.0
35388828              2          5.0
110690689             2          6.5
99525452              2          5.5
...                 ...          ...
58511484              2          5.0
58407519              2          7.5
66450599              2          4.0
14873857              2          4.0
74239933              2          4.5

[7267 rows x 2 columns]

In [7]:
# 2. Choosing a significant level
# For this analysis, the significance level is 5%

significance_level = 0.05
significance_level

0.05

In [8]:
# 3. Finding the p-value

stats.ttest_ind(a=avgFare_amount_credit['fare_amount'], b=avgFare_amount_cash['fare_amount'], equal_var=False)

Ttest_indResult(statistic=6.866800855655372, pvalue=6.797387473030518e-12)

The P-value is 6.80^-12 which is way smaller than 0.05.  Hence, there will be a rejection on the null hypothesis.

<img src="images/Execute.png" width="100" height="100" align=left>

## PACE: **Execute**

Consider the questions in your PACE Strategy Document to reflect on the Execute stage.

### Task 4. Communicate insights with stakeholders

*Ask yourself the following questions:*

1. What business insight(s) can you draw from the result of your hypothesis test?
2. Consider why this A/B test project might not be realistic, and what assumptions had to be made for this educational project.

From the analysis above, we see that more revenue is generated from customers who use credit card compared to customers who use cash.  Based on this observation, the agency can come up with a kind of plan for credit card users such as giving out reduced fare charges after every 10 trips or offering a prestige program for credit card users which could feature services like access to some facilities (health, travel lounge, hotel lounge, etc).

This A/B test project might not be realistic because some of these customers may be visitors who are available for the time period to carry out one or more activities, another is the case whereby there could be some certain scenario where customers were denied paying cash.  Also, there could be other likely explanations thaat may not be accounted for such as a policy placed on the use of credit card based on some threshold amount to be spent.

**Links to documents:**

PACE strategy document: https://docs.google.com/document/d/16e5WgHkbEbPEev2Gh6yI0V6ZM4lsKS3IC04h6AeJmrs/edit?usp=drive_link

Executive summaries: https://docs.google.com/presentation/d/1Ze6Ev2SplXKkBuVBsfPvWD7vYctGD-b-B8yIcJxulW8/edit?usp=drive_link&resourcekey=0-1GL2jPWM0nKV0ZwbSHfm_A

**Congratulations!** You've completed this lab. However, you may not notice a green check mark next to this item on Coursera's platform. Please continue your progress regardless of the check mark. Just click on the "save" icon at the top of this notebook to ensure your work has been logged.